### PR TITLE
Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn

### PR DESCRIPTION
**Pull request by raghuramank100**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37159 Dynamic quantization support for GRU, LSTMCell, RNNCell and GRUCell
* #35961 Fix weight quantization in RNNs
* **#36009 Fix NaN error in dynamic quantization in qLinear, re-enable test_quantized_rnn**

When scale is very small (less than float eps, but greater than minimum double precision value), computation of reciprocal of scale in floating point precision within FBGEMM returns inf, while QuantUtils does not. Changed computation in QuantUtils to occur with floating point precision to re-enable tests.

Differential Revision: [D20853000](https://our.internmc.facebook.com/intern/diff/D20853000/)

### CODE REVIEW

**Comment by jianyuh** [comment]
Still see the errors in the OSS unit test:
https://circleci.com/gh/pytorch/pytorch/5053092?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by z-a-f** [comment]
This change doesn't fix the flakiness of the test. The nan error happens because inside the observer we get `qmin > qmax`, not because of the float. In fact, `float(np.nan)` is a valid NaN value.

**Comment by z-a-f** [comment]
I will take a shot at this issue tomorrow evening, if this is not fixed by then

**Review by z-a-f — CHANGES_REQUESTED** [review]
As mentioned -- this change reenables the test without fixing the flakiness.

File: test/quantization/test_quantization.py [hunk 1]
```diff
@@ -575,7 +575,6 @@ def checkQuantized(model):
         model = quantize_dynamic(NestedModel().eval(), qconfig_dict)
         checkQuantized(model)
 
-    @unittest.skip("temporarily disable the test")
     @given(qengine=st.sampled_from(("fbgemm",)))
```
Comment (supriyar) [inline test/quantization/test_quantization.py:578]: @raghuramank100 could you add `qnnpack` here as well? I believe it was skipped for the same reason.

File: test/quantization/test_quantize.py [hunk 2]
```diff
@@ -624,208 +624,195 @@ def checkQuantized(model):
...
+
+        hx = torch.tensor(h0_vals, dtype=torch.float).unsqueeze(0)
+        cx = torch.tensor(h0_vals, dtype=torch.float).unsqueeze(0)
+
+        if isinstance(ref, torch.nn.LSTM):
+            hiddens = (hx, cx)
+
+        ref_out, ref_hid = ref(x, hiddens)
+
+        for qengine in ['fbgemm', 'qnnpack']:
```
Comment (z-a-f) [inline test/quantization/test_quantize.py:681]: This will fail because `qengine` is not defined

File: test/quantization/test_quantize.py [hunk 3]
```diff
@@ -624,208 +624,195 @@ def checkQuantized(model):
...
+                            out_loaded_packed, hid_loaded_packed = loaded_packed(packed_input, hiddens)
+                            for packed_val, ref_val in zip(out_loaded_packed, ref_out_packed):
+                                if isinstance(packed_val, torch.Tensor):
+                                    torch.testing.assert_allclose(packed_val, ref_val)
+                                else:
+                                    self.assertEqual(packed_val, ref_val)
+
+
+    def test_default_quantized_lstm(self):
+        for qengine in ['fbgemm', 'qnnpack']:
```
Comment (z-a-f) [inline test/quantization/test_quantize.py:788]: Same here

**Review by z-a-f — APPROVED** [review]

File: test/quantization/test_quantize.py [hunk 4]
```diff
@@ -624,208 +624,201 @@ def checkQuantized(model):
...
+            if qengine == 'qnnpack':
+                if IS_PPC or TEST_WITH_UBSAN or IS_MACOS or IS_WINDOWS:
+                    continue
+            with override_quantized_engine(qengine):
+                if qengine in torch.backends.quantized.supported_engines:
```
Comment (z-a-f) [inline test/quantization/test_quantize.py:795]: I think this line should be before the `override_quantized_engine`:

```python
if qengine in torch.backends.quantized.supported_engines:
  continue
with override_quantized_engine(qengine):
```

**Comment by z-a-f** [comment]
Reverting because of the test_qelu breakage https://app.circleci.com/pipelines/github/pytorch/pytorch/160770/workflows/a3ca7391-3ba4-4558-9929-7ff1ffd4794c/

**Pull request closed without merging**